content/2-templating/5-dom-ref/angularRenaissance/inputfocused.component.ts

```diff
@@ -0,0 +1,14 @@
+import { Component, ElementRef, OnInit, viewChild } from "@angular/core";
+
+@Component({
+  standalone: true,
+  selector: "app-inputfocused",
+  template: `<input type="text" #inputRef />`,
+})
+export class InputfocusedComponent implements OnInit {
+  inputRef = viewChild.required<ElementRef<HTMLInputElement>>("inputRef");
+
+  ngOnInit() {
```

Comment:
I don't think so for two reasons:
- the angular team advocates to avoid them as much as possible (since they kinda "break" the pattern)
- that would be a little overkill in order to make only one focus, once

I shouldn't add it in an `ngOnInit` though!
I could either do it in an `ngAfterViewInit` or in a `afterNextRender`

Reply:
It simplifies code and makes it reactive. Maybe in this case where you use static `viewChild()` it doesn't really matter, but if it was `contentChild()`, change to content children would not be detected when you use `ngOnInit()`.
The `effect()` is a great way to merge reactive and non-reactive worlds. Only when you have purely reactive app, you should choose sth like `computed()` instead of `effect()`. Understand the recommendation instead of just following it blindly.
```ts
_ = effect(() => this.inputRef().nativeElement.focus());
```

Reply:
- `ngOnInit` is a mistake here => I copied/pasted the older version and replaced older ways with the new one. About the older way, they used `static: true`  alongside `ngOnInit`  but since the new signal based API don't have it anymore, I just forget to replace it
- `effect`  is like a `tap`  or a `subscribe`  (on the `next`  cb). While it's true its callback could be called only once if the signal changes only once, it could also be called many times. => Why would I use `ngAfterViewInit` or `afterNextRender` here? Because if you don't think about the "how" but about the "what" => you only want to focus once and that's it. If you ever replaced the `viewChild.required` with a `viewChild` within an `@if`, you would probably not expect it to refocus the `input` once again. The effect is not the right tool you would need here.

And no worries, I understand the recomandations pretty well, I only use the tools that fits best to what I'm trying to achieve.

Reply:
Fair enough. Feel free to mark it as resolved.